### ISSUE TITLE
RedisPubSubConnectionImpl.activated() deadlock

### ISSUE THREAD

**Issue by kichik** [mp911de/lettuce#57]
`RedisPubSubConnectionImpl.activated()` can deadlock if another thread is trying to subscribe or unsubscribe at the same time. `RedisPubSubConnectionImpl.subscribe()` calls `RedisAsyncConnectionImpl.dispatch()` which is synchronized. `RedisAsyncConnectionImpl.dispatch()` calls `CommandHandler.write()` which tries to get `writeLock`. `CommandHandler.channelActive()` holds `writeLock` and calls `RedisPubSubConnectionImpl.activated()` which calls `RedisPubSubConnectionImpl.subscribe()` and tries to get the object lock held by the other thread trying to subscribe.

So basically:
* `CommandHandler.channelActive()` holds `writeLock` and tries to get `RedisAsyncConnectionImpl` object lock
* `RedisPubSubConnectionImpl.subscribe()` holds `RedisAsyncConnectionImpl` object lock and then tries to get `writeLock`

**Comment by mp911de** [comment]
I feared, this would happen sooner or later.

**Comment by kichik** [comment]
My solution for this was to execute the subscription in a thread. It can be pushed on the Netty loop too I guess.

**Comment by mp911de** [comment]
I'm wondering whether it is a good idea to introduce a ConnectionOptions class for enabling/disabling reconnects and perhaps how reconnect errors should be handled. This does not solve the problem above but would give a more fine grained control to the user.

**Comment by kichik** [comment]
That will be a good place to add that "ping before thinking the connection is alive" I asked for before. This way the SSL tunnel can be verified before being used.

**Comment by mp911de** [comment]
+1

**Comment by mp911de** [comment]
I think, removing `synchronized` from the `dispatch()` methods should solve the deadlock. The `writeLock` protects the reconnect sequence from other, interfering commands. Clients invoking commands while `channelActive` will simply wait until `activated()` is done.

**Comment by kichik** [comment]
I think there is more reason for this `synchronized`. Another method using it is `close()` and we probably don't want anything writing while that is happening.

**Comment by mp911de** [comment]
`synchronized` on `close()` is important for that case. `close()` should be called rarely in comparison to other methods. The only race which could occur is when using `MULTI` and concurrent threads try to add commands to the `MULTI`. This is in any case a bad idea - with or without `synchronized`. 

I took a look in the older version code. There was no inner lock and the method level `synchronized` on `dispatch()` was intended to prevent races. In certain constellations the commands received responses from earlier commands, therefore the inner lock came into place.

**Comment by mp911de** [comment]
Fixed with #59
